src/io/reader.rs

```diff
@@ -0,0 +1,108 @@
+/// decoding method, and dispatches into the set of supported `ImageDecoder` implementations.
+pub struct Reader<R: Read> {
+    /// The reader.
+    inner: R,
+    /// The format, if one has been set or deduced.
+    format: Option<ImageFormat>,
+}
+
+impl<R: Read> Reader<R> {
+    pub fn new(reader: R) -> Self {
+        Reader {
+            inner: reader,
+            format: None,
+        }
+    }
+
+    /// Construct a reader with specified format.
+    pub fn with_format(reader: R, format: ImageFormat) -> Self {
+        Reader {
+            inner: reader,
+            format: Some(format),
+        }
+    }
+
+    /// Get the currently determined format.
+    pub fn format(&self) -> Option<ImageFormat> {
+        self.format
+    }
+
+    /// Unwrap the reader.
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl Reader<BufReader<File>> {
+    /// Open a file to read. Format will be guessed.
+    /// Note: Failure to guess is not an error. The `Err` is only for `File::open`.
+    pub fn open<P>(path: P) -> io::Result<Self> where P: AsRef<Path> {
+        let path = path.as_ref();
+
+        let file = File::open(path)?;
+        let format = ImageFormat::from_path(path);
+
+        Ok(Reader {
+            inner: BufReader::new(file),
+            format: format.ok(),
+        })
+    }
+}
+
+impl<R: BufRead + Seek> Reader<R> {
+    /// Replace the format with a guess based on the content.
+    ///
+    /// If the guess fails, the format is unchanged. Error is for seek, hence io::Error 
+    /// and not ImageError.
+    pub fn guess_format(&mut self) -> io::Result<()> {
+        let mut start = [0; 16];
+
+        // Save current offset, read start, restore offset.
+        let cur = self.inner.seek(SeekFrom::Current(0))?;
+        let len = io::copy(
+            // Accept shorter files but read at most 16 bytes.
+            &mut self.inner.by_ref().take(16),
+            &mut Cursor::new(&mut start[..]))?;
+        self.inner.seek(SeekFrom::Start(cur))?;
+
+        self.format = free_functions::guess_format(&start[..len as usize]).ok();
+        Ok(())
+    }
+
+    /// Read the image dimensions.
+    /// Uses the current format to construct the correct reader.
+    pub fn image_dimensions(self) -> ImageResult<(u32, u32)> {
+        let format = self.format
+            .map_or_else(|| self.guess_format(), Ok)
+            .map_err(ImageError::IoError)?;
+
+        free_functions::image_dimensions_with_format_impl(self.inner, format)
+    }
+
+    /// Read the image (replaces `load`).
+    /// Uses the current format to construct the correct reader.
+    pub fn load(self) -> ImageResult<DynamicImage> {
+        unimplemented!()
+    }
+}
+
+/// Potential apis, not part of the core proposal:
+impl<R: Read> Reader<R> {
+    /// Load for image formats that don't require seeking.
+    pub fn load_forward(self) -> ImageResult<DynamicImage> {
```

Comment:
I don't think this is required. In almost all cases users should either have a reader that's already seek, or can use Cursor + read_to_end to get one. The rest of the time they can use the ImageDecoder for their specific format. (Which they have to know because none of the format guessing code would work for them...)